src/Futhark/CodeGen/Backends/GenericCSharp/AST.hs

```diff
@@ -0,0 +1,423 @@
+{-# LANGUAGE PostfixOperators #-}
+
+
+module Futhark.CodeGen.Backends.GenericCSharp.AST
+  ( CSExp(..)
+  , CSType(..)
+  , CSComp(..)
+  , CSPrim(..)
+  , CSInt(..)
+  , CSUInt(..)
+  , CSFloat(..)
+  , CSIdx (..)
+  , CSArg (..)
+  , CSStmt(..)
+  , module Language.Futhark.Core
+  , CSProg(..)
+  , CSExcept(..)
+  , CSFunDef(..)
+  , CSFunDefArg
+  , CSClassDef(..)
+  , CSConstructorDef(..)
+  )
+  where
+
+import Language.Futhark.Core
+import Data.List(intersperse)
+import Futhark.Util.Pretty
+import Language.C.Quote.OpenCL()
+import qualified Language.C.Syntax as C
+
+data MemT = Pointer
+          deriving (Eq, Show)
+
+data ArgMemType = ArgOut
+                | ArgRef
+                deriving (Eq, Show)
+
+instance Pretty ArgMemType where
+  ppr ArgOut = text "out"
+  ppr ArgRef = text "ref"
+
+instance Pretty CSComp where
+  ppr (ArrayT t) = ppr t <> text "[]"
+  ppr (TupleT ts) = parens(commasep $ map ppr ts)
+  ppr (SystemTupleT ts) = text "Tuple" <> angles(commasep $ map ppr ts)
+
+data CSInt = Int8T
+           | Int16T
+           | Int32T
+           | Int64T
+           deriving (Eq, Show)
+
+data CSUInt = UInt8T
+            | UInt16T
+            | UInt32T
+            | UInt64T
+            deriving (Eq, Show)
+
+data CSFloat = FloatT
+             | DoubleT
+             deriving (Eq, Show)
+
+data CSType = Composite CSComp
+            | PointerT CSType
+            | Primitive CSPrim
+            | MemoryT String
+            | CLMemoryT String
```

Comment:
Can't you just use an appropriate CustomT instead of CLMemoryT and MemoryT?  The experience we had with the Python codegen was that it's a mistake to add convenience things to the AST that does not map to the real grammar.